src/lsm/grid.zig

```diff
@@ -768,33 +707,102 @@ pub fn GridType(comptime Storage: type) type {
             // of the callback as any nested Grid calls cannot synchronously update the cache).
-            switch (read.callback) {
-                .normal => |callback| callback(read, block.?),
-                .repair => |callback| callback(read, block),
+            read.callback(read, block);
+        }
+
+        /// If the block is not present (or corrupt), we do not attempt to repair it — the repair's
+        /// address/checksum is not assumed to match what "should" be in our block.
+        /// Relatedly we still attempt to read free blocks — they might still hold the requested data.
+        pub fn read_block_repair(
+            grid: *Grid,
+            callback: fn (*Grid.ReadRepair, error{BlockNotFound}!void) void,
+            read: *Grid.ReadRepair,
+            block: BlockPtr,
+            address: u64,
+            checksum: u128,
+        ) void {
+            assert(address > 0);
+
+            assert(grid.superblock.opened);
+            // We try to read the block even when it is free — if we recently released it,
+            // it might be found on disk anyway.
+            maybe(grid.superblock.free_set.is_free(address));
+            // The caller will not attempt to help another replica repair a block that
+            // we are already trying to repair ourselves.
+            assert(!grid.faulty(address, null));
+            maybe(grid.writing(address, null) == .init);
+
+            read.* = .{
+                .callback = callback,
+                .address = address,
+                .checksum = checksum,
+                .block = block,
+                .grid = grid,
+            };
+
+            grid.on_next_tick(read_block_repair_tick_callback, &read.next_tick);
+        }
+
+        fn read_block_repair_tick_callback(next_tick: *Grid.NextTick) void {
+            const read = @fieldParentPtr(ReadRepair, "next_tick", next_tick);
+            const grid = read.grid;
+
+            if (grid.cache.get_index(read.address)) |cache_index| {
+                const cache_block = grid.cache_blocks[cache_index];
+
+                const header = mem.bytesAsValue(vsr.Header, cache_block[0..@sizeOf(vsr.Header)]);
+                assert(header.op == read.address);
+                if (constants.verify) grid.verify_read(read.address, cache_block);
+
+                if (header.checksum == read.checksum) {
+                    stdx.copy_disjoint(.inexact, u8, read.block, cache_block[0..header.size]);
+                    read.callback(read, {});
+                } else {
+                    read.callback(read, error.BlockNotFound);
+                }
+            } else {
+                grid.superblock.storage.read_sectors(
+                    read_block_repair_callback,
+                    &read.completion,
+                    read.block,
+                    .grid,
+                    block_offset(read.address),
+                );
+            }
+        }
```

Comment:
I guess this could be a tiny bit more optimal and avoid introduction of next_tick delay if we are going to rid from storage anyway? 

```
if in_cache {
    memcpy()
    on_next_tick(|| read.callback())
} else {
    read_sectors
}
```

That is, copy data to message synchroneously, and then I think it would be safe to next-tick?